internal/vulnstore/postgres/gc.go

```diff
@@ -0,0 +1,278 @@
+	// all in-flight go routines are guarantee to release their sems.
+	sem.Acquire(context.Background(), cpus)
+
+	close(errC)
+	for e := range errC {
+		if e != nil {
+			return totalOps - deletedOps, fmt.Errorf("received error issuing deletes: %v", err)
+		}
+	}
+	return totalOps - deletedOps, nil
+}
+
+// distinctUpdaters returns all updaters which have registered an update
+// operation.
+func distinctUpdaters(ctx context.Context, pool *pgxpool.Pool) ([]string, error) {
+	const (
+		// will always contain at least two update operations
+		selectUpdaters = `
+SELECT DISTINCT(updater) FROM update_operation;
+`
+	)
+	rows, err := pool.Query(ctx, selectUpdaters)
+	defer rows.Close()
+	if err != nil {
+		return nil, fmt.Errorf("error selecting distinct updaters: %v", err)
+	}
+
+	var updaters []string
+	for rows.Next() {
+		var updater string
+		err := rows.Scan(&updater)
+		switch err {
+		case nil:
+			// hop out
+		default:
+			return nil, fmt.Errorf("error scanning updater: %v", err)
+		}
+		updaters = append(updaters, updater)
+	}
+	if rows.Err() != nil {
+		return nil, rows.Err()
+	}
+	return updaters, nil
+}
+
+// eligibleUpdateOpts returns a list of update operation refs which exceed the specified
+// keep value.
+func eligibleUpdateOpts(ctx context.Context, pool *pgxpool.Pool, keep int) ([]uuid.UUID, int64, error) {
+	const (
+		// this query will return rows of UUID arrays.
+		// each returned array are the UUIDs which exceed the provided keep value
+		updateOps = `
+WITH ordered_ops AS (
+    SELECT array_agg(ref ORDER BY date DESC) AS refs FROM update_operation GROUP BY updater
+)
+SELECT ordered_ops.refs[$1:]
+FROM ordered_ops
+WHERE array_length(ordered_ops.refs, 1) > $2;
+`
+	)
+
+	// gather any update operations exceeding our keep value.
+	// keep+1 is used because PG's array slicing is inclusive,
+	// we want to grab all items once after our keep value.
+	m := []uuid.UUID{}
+	rows, err := pool.Query(ctx, updateOps, keep+1, keep)
+	switch err {
+	case nil:
+	default:
+		return nil, 0, fmt.Errorf("error querying for update operations: %v", err)
+	}
+	defer rows.Close()
+	for rows.Next() {
+		// pgx will not scan directly into a []uuid.UUID
+		tmp := pgtype.UUIDArray{}
+		err := rows.Scan(&tmp)
+		if err != nil {
+			return nil, 0, fmt.Errorf("error scanning update operations: %w", err)
+		}
+		for _, u := range tmp.Elements {
+			m = append(m, u.Bytes) // this works since [16]byte value is assignable to uuid.UUID
+		}
+	}
+	if rows.Err() != nil {
+		return nil, 0, rows.Err()
+	}
+	return m, int64(len(m)), nil
+}
+
+func chunkedCleanup(ctx context.Context, pool *pgxpool.Pool, updater string) error {
+	const (
+		paginatedSelect = `
+SELECT id FROM vuln WHERE vuln.updater = $1 AND id > $2 ORDER BY id ASC LIMIT 10000;
+`
+		shouldDelete = `
+SELECT NOT EXISTS(SELECT 1 FROM uo_vuln WHERE vuln = $1);
+`
+		deleteVuln = `
+DELETE FROM vuln WHERE id = $1;
+`
+	)
+
+	ids := make([]int64, 0, 10000)
+	toDelete := make([]int64, 0, 10000)
```

Comment:
this allocates 160MB per call, it might be better to let the `id` slice get grown as it goes around the loop and have the function at L225 create the batch used in the function at L253.